File: golang/mental1104/mq/factory/factory.go
// Package factory 根据类型安全的后端配置创建公共 Producer 和 Consumer Bridge。
// 本包只负责选择实现，不承载消息发送、消费循环或 SDK 生命周期逻辑。
package factory

import (
	commonmq "github.com/mental1104/common/golang/mental1104/mq"
	kafkamq "github.com/mental1104/common/golang/mental1104/mq/kafka"
	pulsarmq "github.com/mental1104/common/golang/mental1104/mq/pulsar"
)

// NewProducer 根据 config.Backend 创建 Kafka 或 Pulsar ProducerBackend，
// 再用公共 Bridge 包装。Backend 可使用值或非 nil 指针；其他类型返回 ErrorInvalidConfig。
func NewProducer(config commonmq.ProducerConfig) (*commonmq.Producer, error) {
	if config.Backend == nil {
		return nil, commonmq.NewError(commonmq.ErrorInvalidConfig, "create producer", "", "backend config must not be nil", nil)
	}
	var (
		backend commonmq.ProducerBackend
		err     error
	)
	// Factory 只在这里选择具体实现，避免业务代码散落 backend 类型判断。
	switch value := config.Backend.(type) {
	case kafkamq.Config:
		backend, err = kafkamq.NewProducerBackend(config, value)
	case *kafkamq.Config:
		if value == nil {
			err = commonmq.NewError(commonmq.ErrorInvalidConfig, "create producer", commonmq.BackendKafka, "backend config must not be nil", nil)
		} else {
			backend, err = kafkamq.NewProducerBackend(config, *value)
		}
	case pulsarmq.Config:
		backend, err = pulsarmq.NewProducerBackend(config, value)
	case *pulsarmq.Config:
		if value == nil {
			err = commonmq.NewError(commonmq.ErrorInvalidConfig, "create producer", commonmq.BackendPulsar, "backend config must not be nil", nil)
		} else {
			backend, err = pulsarmq.NewProducerBackend(config, *value)
		}
	default:
		err = commonmq.NewError(commonmq.ErrorInvalidConfig, "create producer", config.Backend.BackendType(), "unsupported backend config", nil)
	}
	if err != nil {
		return nil, err
	}
	return commonmq.NewProducer(backend)
}

// NewConsumer 根据 config.Backend 创建 Kafka 或 Pulsar ConsumerBackend，
// 再用公共 Bridge 包装。Backend 可使用值或非 nil 指针；其他类型返回 ErrorInvalidConfig。
func NewConsumer(config commonmq.ConsumerConfig) (*commonmq.Consumer, error) {
	if config.Backend == nil {
		return nil, commonmq.NewError(commonmq.ErrorInvalidConfig, "create consumer", "", "backend config must not be nil", nil)
	}
	var (
		backend commonmq.ConsumerBackend
		err     error
	)
	// Producer 和 Consumer 分别选择小接口，避免形成同时拥有全部能力的上帝 backend。
	switch value := config.Backend.(type) {
	case kafkamq.Config:
		backend, err = kafkamq.NewConsumerBackend(config, value)
	case *kafkamq.Config:
		if value == nil {
			err = commonmq.NewError(commonmq.ErrorInvalidConfig, "create consumer", commonmq.BackendKafka, "backend config must not be nil", nil)
		} else {
			backend, err = kafkamq.NewConsumerBackend(config, *value)
		}
	case pulsarmq.Config:
		backend, err = pulsarmq.NewConsumerBackend(config, value)
	case *pulsarmq.Config:
		if value == nil {
			err = commonmq.NewError(commonmq.ErrorInvalidConfig, "create consumer", commonmq.BackendPulsar, "backend config must not be nil", nil)
		} else {
			backend, err = pulsarmq.NewConsumerBackend(config, *value)
		}
	default:
		err = commonmq.NewError(commonmq.ErrorInvalidConfig, "create consumer", config.Backend.BackendType(), "unsupported backend config", nil)
	}
	if err != nil {
		return nil, err
	}
	return commonmq.NewConsumer(backend)
}

File: cpp/src/mq/factory.cpp
#include "mental1104/mq/factory.h"
#include "mental1104/mq/kafka.h"
#include "mental1104/mq/pulsar.h"

namespace mental1104 {
namespace mq {
namespace {

/// 把 BackendType 转换为稳定日志名称。
std::string backend_name(BackendType type) {
  return type == BackendType::Kafka ? "kafka" : "pulsar";
}

} // namespace

/// 根据多态配置选择 Kafka 或 Pulsar Producer backend。
std::unique_ptr<IProducerBackend>
create_producer_backend(const ProducerConfig &config) {
  if (!config.backend)
    throw MQException(MQError(ErrorCode::InvalidConfig,
                              "create_producer_backend",
                              "backend config must not be null"));
  const std::string name = backend_name(config.backend->backend_type());
  try {
    if (config.backend->backend_type() == BackendType::Kafka) {
      const KafkaBackendConfig *value =
          dynamic_cast<const KafkaBackendConfig *>(config.backend.get());
      if (!value)
        throw MQException(
            MQError(ErrorCode::InvalidConfig, "create_producer_backend",
                    "Kafka backend config has the wrong dynamic type"));
      return create_kafka_producer_backend(config, *value);
    }
    if (config.backend->backend_type() == BackendType::Pulsar) {
      const PulsarBackendConfig *value =
          dynamic_cast<const PulsarBackendConfig *>(config.backend.get());
      if (!value)
        throw MQException(
            MQError(ErrorCode::InvalidConfig, "create_producer_backend",
                    "Pulsar backend config has the wrong dynamic type"));
      return create_pulsar_producer_backend(config, *value);
    }
    throw MQException(MQError(ErrorCode::InvalidConfig,
                              "create_producer_backend",
                              "unsupported backend type"));
  } catch (const MQException &) {
    throw;
  } catch (...) {
    // 第三方 SDK 构造异常在 Factory 边界统一收敛。
    throw MQException(exception_error("create_producer_backend", name));
  }
}

/// 根据多态配置选择 Kafka 或 Pulsar Consumer backend。
std::unique_ptr<IConsumerBackend>
create_consumer_backend(const ConsumerConfig &config) {
  if (!config.backend)
    throw MQException(MQError(ErrorCode::InvalidConfig,
                              "create_consumer_backend",
                              "backend config must not be null"));
  const std::string name = backend_name(config.backend->backend_type());
  try {
    if (config.backend->backend_type() == BackendType::Kafka) {
      const KafkaBackendConfig *value =
          dynamic_cast<const KafkaBackendConfig *>(config.backend.get());
      if (!value)
        throw MQException(
            MQError(ErrorCode::InvalidConfig, "create_consumer_backend",
                    "Kafka backend config has the wrong dynamic type"));
      return create_kafka_consumer_backend(config, *value);
    }
    if (config.backend->backend_type() == BackendType::Pulsar) {
      const PulsarBackendConfig *value =
          dynamic_cast<const PulsarBackendConfig *>(config.backend.get());
      if (!value)
        throw MQException(
            MQError(ErrorCode::InvalidConfig, "create_consumer_backend",
                    "Pulsar backend config has the wrong dynamic type"));
      return create_pulsar_consumer_backend(config, *value);
    }
    throw MQException(MQError(ErrorCode::InvalidConfig,
                              "create_consumer_backend",
                              "unsupported backend type"));
  } catch (const MQException &) {
    throw;
  } catch (...) {
    // 第三方 SDK 构造异常在 Factory 边界统一收敛。
    throw MQException(exception_error("create_consumer_backend", name));
  }
}

/// 创建拥有具体 Producer backend 的 Bridge。
Producer create_producer(const ProducerConfig &config) {
  return Producer(create_producer_backend(config));
}

/// 创建拥有具体 Consumer backend 的 Bridge。
Consumer create_consumer(const ConsumerConfig &config) {
  return Consumer(create_consumer_backend(config));
}

} // namespace mq
} // namespace mental1104
